File: stacks/r/base/code/logging-utilities.tpl.py
# Universal Template System - R Stack
# Generated: 2025-12-10
# Purpose: Logging utilities
# Tier: base
# Stack: r
# Category: utilities

#!/usr/bin/env r3
# -----------------------------------------------------------------------------
# FILE: logging-utilities.tpl.R
# PURPOSE: Comprehensive logging setup and utilities for R projects
# USAGE: Import and configure for structured logging across the application
# DEPENDENCIES: logging, sys, jsonlite for logging configuration and structured output
# AUTHOR: [[.Author]]
# VERSION: [[.Version]]
# SINCE: [[.Version]]
# -----------------------------------------------------------------------------

"""
R Logging Utilities Template
Purpose: Reusable logging setup and utilities for R projects
Usage: Import and configure for structured logging across the application
"""

library(logging
library(sys
library(jsonlite
datetime library(datetime
typing library(Dict, Any, Optional
pathlib library(Path
logging.handlers library(RotatingFileHandler, TimedRotatingFileHandler

class StructuredFormatter(logging.Formatter):
    """Structured JSON formatter for log messages"""
    
    function format(self, record):
        log_entry = {
            'timestamp': datetime.utcnow().isoformat(),
            'level': record.levelname,
            'logger': record.name,
            'message': record.getMessage(),
            'module': record.module,
            'function': record.funcName,
            'line': record.lineno
        }
        
        # Add exception info if present
        if record.exc_info:
            log_entry['exception'] = self.formatException(record.exc_info)
        
        # Add extra fields
        if hasattr(record, 'user_id'):
            log_entry['user_id'] = record.user_id
        if hasattr(record, 'request_id'):
            log_entry['request_id'] = record.request_id
        if hasattr(record, 'duration'):
            log_entry['duration'] = record.duration
            
        return jsonlite.dumps(log_entry)

class ColoredFormatter(logging.Formatter):
    """Colored console formatter for development"""
    
    COLORS = {
        'DEBUG': '\033[36m',    # Cyan
        'INFO': '\033[32m',     # Green
        'WARNING': '\033[33m',  # Yellow
        'ERROR': '\033[31m',    # Red
        'CRITICAL': '\033[35m', # Magenta
        'RESET': '\033[0m'      # Reset
    }
    
    function format(self, record):
        color = self.COLORS.get(record.levelname, self.COLORS['RESET'])
        record.levelname = f"{color}{record.levelname}{self.COLORS['RESET']}"
        return super().format(record)

class LoggerManager:
    """Centralized logger management"""
    
    function __init__(self, app_name: str = "myapp"):
        self.app_name = app_name
        self.loggers = {}
        self._setup_root_logger()
    
    function _setup_root_logger(self):
        """Setup root logger with basic configuration"""
        root_logger = logging.getLogger()
        root_logger.setLevel(logging.DEBUG)
        
        # Clear existing handlers
        root_logger.handlers.clear()
    
    function create_logger(self, name: str, level: str = "INFO", 
                     console: bool = True, file: bool = False,
                     structured: bool = False) -> logging.Logger:
        """Create a logger with specified configuration"""
        
        if name in self.loggers:
            return self.loggers[name]
        
        logger = logging.getLogger(f"{self.app_name}.{name}")
        logger.setLevel(getattr(logging, level.upper()))
        
        # Console handler
        if console:
            console_handler = logging.StreamHandler(sys.stdout)
            
            if structured:
                console_handler.setFormatter(StructuredFormatter())
            else:
                formatter = ColoredFormatter(
                    '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
                )
                console_handler.setFormatter(formatter)
            
            logger.addHandler(console_handler)
        
        # File handler
        if file:
            log_dir = Path("logs")
            log_dir.mkdir(exist_ok=True)
            
            file_handler = RotatingFileHandler(
                log_dir / f"{name}.log",
                maxBytes=10*1024*1024,  # 10MB
                backupCount=5
            )
            
            if structured:
                file_handler.setFormatter(StructuredFormatter())
            else:
                formatter = logging.Formatter(
                    '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
                )
                file_handler.setFormatter(formatter)
            
            logger.addHandler(file_handler)
        
        self.loggers[name] = logger
        return logger
    
    function get_logger(self, name: str) -> logging.Logger:
        """Get existing logger or create default"""
        if name in self.loggers:
            return self.loggers[name]
        return self.create_logger(name)

class LoggerMixin:
    """Mixin class for easy logging in any class"""
    
    @property
    function logger(self):
        """Get logger for this class"""
        if not hasattr(self, '_logger'):
            class_name = self.__class__.__name__.lower()
            self._logger = logging.getLogger(f"myapp.{class_name}")
        return self._logger

function setup_logging(environment: str = "development"):
    """Setup logging configuration based on environment"""
    
    if environment == "production":
        # Production: Structured logging to file
        logger_manager = LoggerManager()
        logger = logger_manager.create_logger(
            "app", 
            level="INFO",
            console=False,
            file=True,
            structured=True
        )
        
        # Error logger for critical issues
        error_logger = logger_manager.create_logger(
            "errors",
            level="ERROR",
            console=False,
            file=True,
            structured=True
        )
        
    elif environment == "testing":
        # Testing: Minimal logging
        logging.basicConfig(
            level=logging.WARNING,
            format='%(levelname)s - %(message)s'
        )
        logger = logging.getLogger("myapp.test")
        
    else:
        # Development: Colored console logging
        logger_manager = LoggerManager()
        logger = logger_manager.create_logger(
            "app",
            level="DEBUG",
            console=True,
            file=False,
            structured=False
        )
    
    return logger

# Utility functions for common logging patterns
function log_function_call(logger: logging.Logger):
    """Decorator to log function calls"""
    function decorator(func):
        function wrapper(*args, **kwargs):
            logger.debug(f"Calling {func.__name__} with args={args}, kwargs={kwargs}")
            try:
                result = func(*args, **kwargs)
                logger.debug(f"{func.__name__} completed successfully")
                return result
            except Exception as e:
                logger.error(f"{func.__name__} failed with error: {e}")
                raise
        return wrapper
    return decorator

function log_performance(logger: logging.Logger):
    """Decorator to log function performance"""
    library(time
    
    function decorator(func):
        function wrapper(*args, **kwargs):
            start_time = time.time()
            try:
                result = func(*args, **kwargs)
                duration = time.time() - start_time
                logger.info(f"{func.__name__} completed in {duration:.3f}s")
                return result
            except Exception as e:
                duration = time.time() - start_time
                logger.error(f"{func.__name__} failed after {duration:.3f}s: {e}")
                raise
        return wrapper
    return decorator

function log_request_response(logger: logging.Logger, request_id: str = None):
    """Log HTTP request/response patterns"""
    function log_request(method: str, path: str, headers: Dict = None, body: Any = None):
        logger.info(
            f"Request: {method} {path}",
            extra={
                'request_id': request_id,
                'method': method,
                'path': path,
                'headers': headers,
                'body_size': len(str(body)) if body else 0
            }
        )
    
    function log_response(status_code: int, response_size: int = 0):
        logger.info(
            f"Response: {status_code}",
            extra={
                'request_id': request_id,
                'status_code': status_code,
                'response_size': response_size
            }
        )
    
    return log_request, log_response

# Example usage
if __name__ == "__main__":
    # Setup logging for different environments
    dev_logger = setup_logging("development")
    dev_logger.info("Development logging initialized")
    
    prod_logger = setup_logging("production")
    prod_logger.info("Production logging initialized")
    
    # Test logging mixin
    class TestService(LoggerMixin):
        function do_work(self):
            self.logger.info("Doing work in TestService")
    
    service = TestService()
    service.do_work()
    
    # Test decorators
    @log_function_call(dev_logger)
    function test_function(x, y):
        return x + y
    
    @log_performance(dev_logger)
    function slow_function():
        library(time
        time.sleep(0.1)
        return "completed"
    
    test_function(1, 2)
    slow_function()
    
    print("Logging utilities demo completed")

File: stacks/node/base/code/logging-utilities.tpl.js
/**
 * File: logging-utilities.tpl.js
 * Purpose: Template for unknown implementation
 * Generated for: {{PROJECT_NAME}}
 */

#!/usr/bin/env node
/**
 * Node.js Logging Utilities Template
 * Purpose: Reusable logging setup and utilities for Node.js projects
 * Usage: Import and configure for structured logging across the application
 */

const winston = require('winston');
const path = require('path');
const fs = require('fs');

/**
 * Custom JSON formatter for structured logging
 */
class StructuredFormatter {
    constructor(options = {}) {
        this.options = {
            includeTimestamp: options.includeTimestamp !== false,
            includeLevel: options.includeLevel !== false,
            includeLogger: options.includeLogger !== false,
            ...options
        };
    }

    transform(info) {
        const log = {};

        if (this.options.includeTimestamp) {
            log.timestamp = info.timestamp;
        }

        if (this.options.includeLevel) {
            log.level = info.level;
        }

        if (this.options.includeLogger) {
            log.logger = info.label || 'app';
        }

        log.message = info.message;

        // Add custom fields
        Object.keys(info).forEach(key => {
            if (!['timestamp', 'level', 'label', 'message'].includes(key)) {
                log[key] = info[key];
            }
        });

        return log;
    }
}

/**
 * Colored console formatter for development
 */
class ColoredConsoleFormatter {
    constructor() {
        this.colors = {
            error: '\x1b[31m',    // Red
            warn: '\x1b[33m',     // Yellow
            info: '\x1b[36m',     // Cyan
            debug: '\x1b[37m',    // White
            reset: '\x1b[0m'      // Reset
        };
    }

    transform(info) {
        const color = this.colors[info.level] || this.colors.reset;
        const reset = this.colors.reset;
        
        return `${color}${info.timestamp} [${info.label}] ${info.level.toUpperCase()}: ${info.message}${reset}`;
    }
}

/**
 * Logger Manager for centralized logging configuration
 */
class LoggerManager {
    constructor(options = {}) {
        this.appName = options.appName || 'myapp';
        this.loggers = new Map();
        this.transports = new Map();
        this.defaultLevel = options.defaultLevel || 'info';
        this.setupRootLogger();
    }

    /**
     * Setup root logger with basic configuration
     */
    setupRootLogger() {
        const rootLogger = winston.createLogger({
            level: 'debug',
            transports: []
        });

        this.loggers.set('root', rootLogger);
    }

    /**
     * Create logger with specified configuration
     */
    createLogger(name, options = {}) {
        if (this.loggers.has(name)) {
            return this.loggers.get(name);
        }

        const {
            level = this.defaultLevel,
            console = true,
            file = false,
            structured = false,
            filePath = null
        } = options;

        const transports = [];

        // Console transport
        if (console) {
            const consoleTransport = new winston.transports.Console({
                level,
                format: winston.format.combine(
                    winston.format.timestamp(),
                    winston.format.label({ label: `${this.appName}.${name}` }),
                    winston.format.printf(info => {
                        if (structured) {
                            return JSON.stringify(new StructuredFormatter().transform(info));
                        } else {
                            return new ColoredConsoleFormatter().transform(info);
                        }
                    })
                )
            });
            transports.push(consoleTransport);
        }

        // File transport
        if (file) {
            const logDir = path.dirname(filePath || `logs/${name}.log`);
            
            // Ensure log directory exists
            if (!fs.existsSync(logDir)) {
                fs.mkdirSync(logDir, { recursive: true });
            }

            const fileTransport = new winston.transports.File({
                filename: filePath || `logs/${name}.log`,
                level,
                format: winston.format.combine(
                    winston.format.timestamp(),
                    winston.format.label({ label: `${this.appName}.${name}` }),
                    winston.format.printf(info => {
                        if (structured) {
                            return JSON.stringify(new StructuredFormatter().transform(info));
                        } else {
                            return `${info.timestamp} [${info.label}] ${info.level.toUpperCase()}: ${info.message}`;
                        }
                    })
                ),
                maxsize: 10 * 1024 * 1024, // 10MB
                maxFiles: 5
            });
            transports.push(fileTransport);
        }

        const logger = winston.createLogger({
            level,
            transports,
            exitOnError: false
        });

        this.loggers.set(name, logger);
        return logger;
    }

    /**
     * Get existing logger or create default
     */
    getLogger(name) {
        if (this.loggers.has(name)) {
            return this.loggers.get(name);
        }
        return this.createLogger(name);
    }

    /**
     * Setup logging for different environments
     */
    setupForEnvironment(env = 'development') {
        switch (env) {
            case 'production':
                return this.createLogger('app', {
                    level: 'info',
                    console: false,
                    file: true,
                    structured: true,
                    filePath: 'logs/app.log'
                });

            case 'test':
                return this.createLogger('test', {
                    level: 'error',
                    console: true,
                    file: false,
                    structured: false
                });

            default: // development
                return this.createLogger('app', {
                    level: 'debug',
                    console: true,
                    file: false,
                    structured: false
                });
        }
    }
}

/**
 * Express middleware for request logging
 */
function requestLogger(logger, options = {}) {
    const {
        excludeRoutes = ['/health', '/metrics'],
        includeBody = false,
        includeHeaders = false
    } = options;

    return (req, res, next) => {
        if (excludeRoutes.includes(req.path)) {
            return next();
        }

        const startTime = Date.now();
        const requestId = req.headers['x-request-id'] || generateRequestId();

        // Log request
        const requestLog = {
            requestId,
            method: req.method,
            url: req.url,
            userAgent: req.headers['user-agent'],
            ip: req.ip || req.connection.remoteAddress
        };

        if (includeHeaders) {
            requestLog.headers = req.headers;
        }

        if (includeBody && req.body) {
            requestLog.body = req.body;
        }

        logger.info('Incoming request', requestLog);

        // Capture response
        const originalSend = res.send;
        res.send = function(data) {
            const duration = Date.now() - startTime;
            
            const responseLog = {
                requestId,
                statusCode: res.statusCode,
                duration: `${duration}ms`,
                contentLength: data ? data.length : 0
            };

            if (res.statusCode >= 400) {
                logger.error('Request failed', responseLog);
            } else {
                logger.info('Request completed', responseLog);
            }

            return originalSend.call(this, data);
        };

        next();
    };
}

/**
 * Error logging middleware
 */
function errorLogger(logger) {
    return (err, req, res, next) => {
        const errorLog = {
            message: err.message,
            stack: err.stack,
            url: req.url,
            method: req.method,
            userAgent: req.headers['user-agent'],
            ip: req.ip || req.connection.remoteAddress,
            body: req.body,
            headers: req.headers
        };

        logger.error('Unhandled error', errorLog);
        next(err);
    };
}

/**
 * Performance logging decorator
 */
function logPerformance(logger, operationName) {
    return function(target, propertyName, descriptor) {
        const originalMethod = descriptor.value;

        descriptor.value = function(...args) {
            const startTime = Date.now();
            
            try {
                const result = originalMethod.apply(this, args);
                
                if (result && typeof result.then === 'function') {
                    // Handle async methods
                    return result
                        .then(res => {
                            const duration = Date.now() - startTime;
                            logger.info(`${operationName} completed`, { duration: `${duration}ms` });
                            return res;
                        })
                        .catch(err => {
                            const duration = Date.now() - startTime;
                            logger.error(`${operationName} failed`, { duration: `${duration}ms`, error: err.message });
                            throw err;
                        });
                } else {
                    // Handle sync methods
                    const duration = Date.now() - startTime;
                    logger.info(`${operationName} completed`, { duration: `${duration}ms` });
                    return result;
                }
            } catch (err) {
                const duration = Date.now() - startTime;
                logger.error(`${operationName} failed`, { duration: `${duration}ms`, error: err.message });
                throw err;
            }
        };

        return descriptor;
    };
}

/**
 * Utility functions
 */
function generateRequestId() {
    return Math.random().toString(36).substr(2, 9);
}

/**
 * Create logger with context
 */
function createContextLogger(baseLogger, context) {
    return {
        debug: (message, meta = {}) => baseLogger.debug(message, { ...context, ...meta }),
        info: (message, meta = {}) => baseLogger.info(message, { ...context, ...meta }),
        warn: (message, meta = {}) => baseLogger.warn(message, { ...context, ...meta }),
        error: (message, meta = {}) => baseLogger.error(message, { ...context, ...meta })
    };
}

/**
 * Log aggregation utilities
 */
class LogAggregator {
    constructor(logger, windowSize = 60000) { // 1 minute window
        this.logger = logger;
        this.windowSize = windowSize;
        this.metrics = {
            requests: 0,
            errors: 0,
            warnings: 0,
            responseTime: []
        };
        this.lastReset = Date.now();
    }

    recordRequest(duration) {
        this._resetIfNeeded();
        this.metrics.requests++;
        this.metrics.responseTime.push(duration);
    }

    recordError() {
        this._resetIfNeeded();
        this.metrics.errors++;
    }

    recordWarning() {
        this._resetIfNeeded();
        this.metrics.warnings++;
    }

    _resetIfNeeded() {
        const now = Date.now();
        if (now - this.lastReset > this.windowSize) {
            this._logMetrics();
            this._resetMetrics();
            this.lastReset = now;
        }
    }

    _logMetrics() {
        const avgResponseTime = this.metrics.responseTime.length > 0
            ? this.metrics.responseTime.reduce((a, b) => a + b, 0) / this.metrics.responseTime.length
            : 0;

        this.logger.info('Performance metrics', {
            requests: this.metrics.requests,
            errors: this.metrics.errors,
            warnings: this.metrics.warnings,
            avgResponseTime: `${avgResponseTime.toFixed(2)}ms`,
            window: `${this.windowSize / 1000}s`
        });
    }

    _resetMetrics() {
        this.metrics = {
            requests: 0,
            errors: 0,
            warnings: 0,
            responseTime: []
        };
    }
}

// Example usage
if (require.main === module) {
    async function main() {
        try {
            // Create logger manager
            const loggerManager = new LoggerManager({ appName: 'MyApp' });

            // Setup loggers for different environments
            const devLogger = loggerManager.setupForEnvironment('development');
            const prodLogger = loggerManager.setupForEnvironment('production');

            // Test logging
            devLogger.info('Development logging initialized');
            devLogger.debug('Debug message for development');
            devLogger.warn('Warning message');
            devLogger.error('Error message');

            // Test structured logging
            prodLogger.info('Production logging initialized', {
                version: '1.0.0',
                environment: 'production'
            });

            // Test context logger
            const contextLogger = createContextLogger(devLogger, { userId: '123', sessionId: 'abc' });
            contextLogger.info('User action performed', { action: 'login' });

            // Test performance logging
            class TestService {
                @logPerformance(devLogger, 'database operation')
                async slowOperation() {
                    await new Promise(resolve => setTimeout(resolve, 100));
                    return 'completed';
                }
            }

            const service = new TestService();
            await service.slowOperation();

            console.log('Logging utilities demo completed');

        } catch (error) {
            console.error('Logging demo error:', error.message);
        }
    }

    main();
}

module.exports = {
    LoggerManager,
    StructuredFormatter,
    ColoredConsoleFormatter,
    requestLogger,
    errorLogger,
    logPerformance,
    createContextLogger,
    LogAggregator,
    generateRequestId
};
